apps/prairielearn/src/ee/models/enrollment.ts

```diff
@@ -46,3 +50,75 @@ export async function getEnrollmentCountsForCourseInstance(
     free: result?.free ?? 0,
   };
 }
+
+/**
+ * Performs enterprise-specific checks for a potential enrollment:
+ *
+ * - The new enrollment must not push either the institution or the course instance
+ *   over either of their enrollment limits.
+ * - The user must have any necessary plan grants.
+ *
+ * This function will return `true` if the enrollment would be allowed.
+ * Otherwise, it will redirect the user to the appropriate page and
+ * return `false`.
+ */
+export async function checkEnterpriseEnrollment(
+  res: Response,
```

Comment:
I've generally tried to avoid passing `req` or `res` into library and model functions. Instead I try to have the caller look at the return value and do the redirect. I prefer this so that lib/model functions stay as pure as possible (not taking/modifying huge objects like req/res/locals). I acknowledge that in this case it may be a bit tricky with the two different possible redirects, so we'd need to return a value describing what the problem was so the caller knew how to redirect.

Reply:
I can do that, but this function is itself used by `insertCheckedEnrollment` in `src/models/enrollment.ts`. `insertCheckedEnrollment`, in turn, is used in both the new auto-enroll middleware and the enrollment page, and it feels bad to me to reproduce the redirection logic in both places. Would you prefer a `maybeRedirectForPotentialEnterpriseEnrollmentStatus` function that encapsulates *just* the redirection logic, and then call that anywhere we want to make an enrollment?

Reply:
I did the first half of this in 4ccf5ba. I still have `insertCheckedEnrollment` directly redirecting; let's talk about the best way to do that.

Reply:
I made the second part of this change in d85d3af: instead of passing all of `res`, we now only pass `res.redirect`. I think this works pretty well!